test/grouping.jl

```diff
@@ -210,10 +210,13 @@ module TestGrouping
             @test vcat(v[1], v[2], v[3], v[4]) == by(f8, df, cols, sort=sort)
         end
 
-        # testing pool overflow
-        df2 = DataFrame(v1 = categorical(collect(1:1000)), v2 = categorical(fill(1, 1000)))
-        @test groupby(df2, [:v1, :v2]).starts == collect(1:1000)
-        @test groupby(df2, [:v2, :v1]).starts == collect(1:1000)
+        # test number of potential combinations higher than typemax(Int32)
+        N = 2000
+        df2 = DataFrame(v1 = levels!(categorical(rand(1:N, 100)), collect(1:N)),
+                        v2 = levels!(categorical(rand(1:N, 100)), collect(1:N)),
+                        v3 = levels!(categorical(rand(1:N, 100)), collect(1:N)))
+        df2b = mapcols(x -> convert(Vector{Int}, x), df2)
+        @test groupby(df2, [:v1, :v2, :v3]) == groupby(df2b, [:v1, :v2, :v3])
```

Comment:
Very good point, that actually uncovered a bug when sorting and when skipping missing introduced by https://github.com/JuliaData/DataFrames.jl/pull/1651. It's actually even more tricky, because the internal fields can be different, yet the result be equivalent (due to permutations). But AFAICT it's enough to check for equality and for equality of `groups` fields (which are redundant, and stored only for performance).

Reply:
Yes, they should be covered now.

Reply:
Actually I've found an even stricter way of checking this: each time we call `groupby`, ensure fields are consistent. Let me know what you think.

Reply:
Looks good. Thank you.

What I also meant that comparison of two `GroupedDataFrame`s is OK for the user (we compare values), but for the tests we should have a method that checks if actually we select exactly the same rows in both cases. Otherwise if you would do something like `groupby(DataFrame(a=[1,2,1,2], b=1), :a)` then all tests would pass even inf the `groupby` method were wrong (as `b` column is constant so it will always compare as `true` even if we take rows in an inconsistent order to `SubDataFrames`). Is it clear what I mean? The inconsistency would become apparent if we later written e.g. `b=[1,2,3,4]` in parent.

An alternative is to always use some `id` column in every `DataFrame` in our tests which has values from `1` to `nrows(df)` sorted. Then what we have now is guaranteed to catch such problems.

Reply:
Indeed. After checking, it appears that all tests use unique (or random) values for a non-grouping column, so that should be OK.